### DISCUSSION TITLE
Most performant way to render large number of PNGs

### DISCUSSION THREAD

**Discussion by ameyms** [pixijs/pixijs#9716] [Q&A]
👋🏾 Hello!

_This might be a n00b question so apologies in advance._

# Content
I am trying to render a graph (with nodes and edges). There could potentially be tens of thousands of nodes in the graph at any given point in time.

There are 5-7 of types of nodes and each **type of node** has a distinct icon. So basically, there are 5-7 different icons I want to render, one in each node.

I have currently chosen to use PNG to render these icons.

# Problem

The graph renders fine without nodes (just using `Graphics` and other PixiJS APIs like `drawRoundedRect`). It scales to 70K nodes quite well.

However, if I add icons, the perf is miserable even with a 1000 nodes.
What can I do differently to improve perf?

# Code Snippet

```typescript
for (let index = 0; index < 5; index++) { // here's where Im controlling  how many nodes to add just for testing
            const kind = NODE_KINDS[index % 5]; 
            const iconGraphics = new Graphics()
                .beginTextureFill({ texture: this.iconTextures[kind] })
                .drawCircle(ICON_SIZE / 2, ICON_SIZE / 2, ICON_SIZE / 2)
                .endFill();

            const newX = index * 120 + 120;
            const newY = 120;

            const node = new Graphics(NODE_TEMPLATE.geometry);

            node.position.x = newX;
            node.position.y = newY;
            nodes.push(node);

            const icon = new Graphics(iconGraphics.geometry);

            icon.position.x = newX + NODE_SIZE / 2 - ICON_SIZE / 2;
            icon.position.y = newY + NODE_SIZE / 2 - ICON_SIZE / 2;
            nodes.push(icon);

            const label = new Text(
                kind.toUpperCase(),
                new TextStyle({
                    fontSize: 48,
                    fontFamily: 'Inter',
                    align: 'center',
                })
            );

            label.position.x = newX;
            label.position.y = newY + NODE_SIZE + 16;
            if (x != null && y != null) {
                const edge = new Graphics()
                    .lineStyle(2, 0x66_66_66)
                    .moveTo(x + NODE_SIZE / 2, y + NODE_SIZE / 2)
                    .lineTo(newX + NODE_SIZE / 2, newY + NODE_SIZE / 2);

                this.root.addChild(edge);
            }
            nodes.push(label);

            x = newX;
            y = newY;
        }

        for (const n of nodes) this.root.addChild(n);
        this.view.stage.addChild(this.root);
```

**Comment by RecursiveVoid** [comment]
To render large number of PNGs, You can use **cacheAsBitmap** functionality of the **container**. 
Add all the PNGs as child in one container and set **cacheAsBitmap** to **True**  

https://pixijs.com/examples/basic/cache-as-bitmap

**Comment by zszq** [comment]
Glad to see this question, sorry I can't solve your problem. But I also want to use pixi.js to implement the same thing as you to improve my pixi.js skills, but how did you achieve 70k nodes. Can I check your source code? I want to know how you implemented it?
